File: tailor/listeners/constantdeclistener.py
from tailor.swift.swiftlistener import SwiftListener
from tailor.swift.swiftparser import SwiftParser
from tailor.utils import charformat


class ConstantDecListener(SwiftListener):

    def __init__(self, printer):
        self.__printer = printer

    def enterIdentifier(self, ctx):
        constant_name = ctx.getText()
        constant_decl_ctx = self.__get_constant_declaration(ctx)
        if (self.__is_global(constant_decl_ctx)
                or self.__inside_class(constant_decl_ctx)
                or self.__inside_struct(constant_decl_ctx)):
            if not (charformat.is_upper_camel_case(constant_name)
                    or charformat.is_lower_camel_case(constant_name)):
                self.__printer.error('Global constant should be either' +
                                     ' lowerCamelCase or UpperCamelCase',
                                     ctx=ctx)
        else:
            if not charformat.is_lower_camel_case(constant_name):
                self.__printer.error('Constant should be in lowerCamelCase',
                                     ctx=ctx)

    def __get_constant_declaration(self, ctx):
        if ctx.parentCtx:
            if isinstance(ctx.parentCtx,
                          SwiftParser.ConstantDeclarationContext):
                return ctx.parentCtx
            else:
                return self.__get_constant_declaration(ctx.parentCtx)
        else:
            return None

    def __is_global(self, ctx):
        parent = self.__get_nth_parent(ctx, 3)
        return parent and isinstance(parent, SwiftParser.TopLevelContext)

    def __inside_class(self, ctx):
        root_ctx = self.__get_nth_parent(self.__get_declaration_root(ctx), 1)
        return root_ctx and isinstance(root_ctx,
                                       SwiftParser.ClassDeclarationContext)

    def __inside_struct(self, ctx):
        root_ctx = self.__get_nth_parent(self.__get_declaration_root(ctx), 1)
        return root_ctx and isinstance(root_ctx,
                                       SwiftParser.StructDeclarationContext)

    @staticmethod
    # pylint: disable=invalid-name
    def __get_nth_parent(ctx, n):
        if ctx is None:
            return None
        while n != 0:
            n -= 1
            ctx = ctx.parentCtx
            if ctx is None:
                return None
        return ctx

    @staticmethod
    def __get_declaration_root(ctx):
        if ctx is None:
            return None
        while True:
            ctx = ctx.parentCtx
            if not (isinstance(ctx, SwiftParser.DeclarationsContext)
                    or isinstance(ctx, SwiftParser.DeclarationContext)):
                break
        return ctx

File: src/main/java/com/sleekbyte/tailor/listeners/ConstantDecListener.java
package com.sleekbyte.tailor.listeners;

import com.sleekbyte.tailor.antlr.SwiftBaseListener;
import com.sleekbyte.tailor.antlr.SwiftParser;
import com.sleekbyte.tailor.common.Location;
import com.sleekbyte.tailor.common.Messages;
import com.sleekbyte.tailor.common.Rules;
import com.sleekbyte.tailor.utils.CharFormatUtil;
import com.sleekbyte.tailor.utils.ListenerUtil;
import com.sleekbyte.tailor.utils.ParseTreeUtil;
import org.antlr.v4.runtime.ParserRuleContext;

/**
 * Listener for constant declarations.
 */
public class ConstantDecListener extends SwiftBaseListener {

    private ConstructListener listener;

    /**
     * Creates a ConstantDecListener object and saves a ConstructListener object.
     */
    ConstantDecListener(ConstructListener listener) {
        this.listener = listener;
    }

    @Override
    public void enterIdentifier(SwiftParser.IdentifierContext ctx) {
        String constantName = ctx.getText();
        ParserRuleContext constantDecContext = getConstantDeclaration(ctx);
        Location location = ListenerUtil.getContextStartLocation(ctx);

        if (isGlobal(constantDecContext) || insideClass(constantDecContext) || insideStruct(constantDecContext)) {
            if (listener.ruleEnabled(Rules.CONSTANT_NAMING)
                && !CharFormatUtil.isUpperCamelCase(constantName)
                && !CharFormatUtil.isLowerCamelCase(constantName)) {

                listener.printer.error(Messages.GLOBAL + Messages.CONSTANT + Messages.GLOBAL_CONSTANT_NAMING, location);
            } else if (listener.ruleEnabled(Rules.K_PREFIXED) && CharFormatUtil.isKPrefixed(constantName)) {
                listener.printer.warn(Messages.CONSTANT + Messages.NAME + Messages.K_PREFIXED, location);
            }
        } else {
            if (listener.ruleEnabled(Rules.CONSTANT_NAMING) && !CharFormatUtil.isLowerCamelCase(constantName)) {
                listener.printer.error(Messages.CONSTANT + Messages.LOWER_CAMEL_CASE, location);
            } else if (listener.ruleEnabled(Rules.K_PREFIXED) && CharFormatUtil.isKPrefixed(constantName)) {
                listener.printer.warn(Messages.CONSTANT + Messages.NAME + Messages.K_PREFIXED, location);
            }
        }

        (new MaxLengthVerifier(listener.printer)).verifyNameLength(Messages.CONSTANT + Messages.NAME,
            listener.maxLengths.maxNameLength, ctx);
    }

    private ParserRuleContext getConstantDeclaration(ParserRuleContext ctx) {
        if (ctx.getParent() != null) {
            if (ctx.getParent() instanceof SwiftParser.ConstantDeclarationContext) {
                return ctx.getParent();
            } else {
                return getConstantDeclaration(ctx.getParent());
            }
        } else {
            return null;
        }
    }

    private boolean isGlobal(ParserRuleContext ctx) {
        ParserRuleContext parentCtx = ParseTreeUtil.getNthParent(ctx, 3);
        return parentCtx != null && (parentCtx instanceof SwiftParser.TopLevelContext);
    }

    private boolean insideClass(ParserRuleContext ctx) {
        ParserRuleContext rootCtx = ParseTreeUtil.getNthParent(ConstantDecListener.getDeclarationRoot(ctx), 1);
        return rootCtx != null && (rootCtx instanceof SwiftParser.ClassDeclarationContext);
    }

    private boolean insideStruct(ParserRuleContext ctx) {
        ParserRuleContext rootCtx = ParseTreeUtil.getNthParent(ConstantDecListener.getDeclarationRoot(ctx), 1);
        return rootCtx != null && (rootCtx instanceof SwiftParser.StructDeclarationContext);
    }

    private static ParserRuleContext getDeclarationRoot(ParserRuleContext ctx) {
        if (ctx == null) {
            return null;
        }
        while (true) {
            ctx = ctx.getParent();
            if (!(ctx instanceof SwiftParser.DeclarationsContext)
                && !(ctx instanceof SwiftParser.DeclarationContext)) {
                break;
            }
        }
        return ctx;
    }

}
